include/seqan3/argument_parser/validators.hpp

```diff
@@ -297,68 +285,350 @@ class file_ext_validator
+        std::string tmp_str = path.extension().string();
+        auto drop_less_ext = tmp_str | view::drop(1);
+
+        // Compares the extensions in lower case.
+        auto cmp_lambda = [&] (std::string const & cmp)
+        {
+            return std::ranges::equal(cmp | view::to_lower, drop_less_ext | view::to_lower);
+        };
+
+        // Check if requested extension is present.
+        if (std::ranges::find_if(extensions, cmp_lambda) == extensions.end())
+        {
+            throw parser_invalid_argument{detail::to_string("Expected one of the following valid extensions: ",
+                                                             extensions, "! Got ", drop_less_ext, " instead!")};
+        }
+    }
+
+    /*!\brief Checks if the given path is readable.
+     * \param path The path to check.
+     * \returns `true` if readable, otherwise `false`.
+     */
+    bool path_is_readable(std::filesystem::path const & path) const
+    {
+        // Check if input directory is a readable.
+        if (std::filesystem::is_directory(path))
+        {
+            std::error_code ec{};
+            std::filesystem::directory_iterator{path, ec};  // if directory iterator cannot be created, ec will be set.
+            return !static_cast<bool>(ec); // test the error code.
+        }
+
+        // Must be a regular file.
+        if (!std::filesystem::is_regular_file(path))
+            throw std::invalid_argument{detail::to_string("Expected a regular file!", path, "!")};
+
+        std::ifstream file{path};
+        return file.is_open() && file.good();
+    }
+
+    /*!\brief Checks if the given path is writable.
+     * \param path The path to check.
+     * \returns `true` if writable, otherwise `false`.
+     */
+    bool filename_is_writable(std::filesystem::path const & path) const
+    {
+        std::ofstream file{path};
+        bool is_open = file.is_open();
+        bool is_good = file.good();
+        file.close();
+
+        // Try to remove file.
+        std::filesystem::remove(path);
+
+        return is_open && is_good;
+    }
+
+    //!\brief Stores the extensions.
+    std::vector<std::string> extensions{};
+};
+
+/*!\brief A validator that checks if a given path is a valid input file.
+ * \ingroup argument_parser
+ * \implements seqan3::validator_concept
+ *
+ * \details
+ *
+ * On construction, the validator can receive a list (std::vector over std::string) of valid file extensions.
+ * The struct acts as a functor that throws a seqan3::parser_invalid_argument exception whenever a given filename's
+ * extension (sts::filesystem::path) is not in the given list of valid file extensions, if the file does not exist, or
+ * if the file does not have the proper read permissions.
+ *
+ * \snippet test/snippet/argument_parser/validators_input_file.cpp usage
+ *
+ * \note The validator works on every type that can be implicitly converted to std::filesystem::path.
+ */
+class input_file_validator : public file_validator_base
```

Comment:
Shall it be possible that other validators derive further from this class?

* If yes, please mark the same functions as virtual as in the base class (destructor, operator()).
* If no, please mark this class and/or the respective virtual functions as `final`.